### ISSUE TITLE
No more monitoring info when passive becomes active

### ISSUE THREAD

**Issue by mathieucarbou** [Terracotta-OSS/terracotta-core#409]
I have a stack trace because entities are recreated BEFORE the platform re-create the monitoring tree when the server becomes active

```
[puts_can_be_seen_on_other_clients_after_failover(org.terracotta.management.integration.tests.FailoverIT)][testServer0] 2016-12-19 16:12:09,330 [WorkerThread(l2_election_handler_stage, 0)] INFO com.tc.l2.state.StateManagerImpl - Becoming State[ ACTIVE-COORDINATOR ]
[puts_can_be_seen_on_other_clients_after_failover(org.terracotta.management.integration.tests.FailoverIT)][testServer0] 2016-12-19 16:12:09,330 [WorkerThread(l2_election_handler_stage, 0)] INFO com.terracottatech.console - Becoming State[ ACTIVE-COORDINATOR ]
[puts_can_be_seen_on_other_clients_after_failover(org.terracotta.management.integration.tests.FailoverIT)][testServer0] 2016-12-19 16:12:09,331 [WorkerThread(l2_election_handler_stage, 0)] INFO com.tc.l2.state.ElectionManagerImpl - Declared as Winner: Winner is : Enrollment [ NodeID[localhost:22965], isNew = false, weights = 11,0,1968,-6114697689862609498 ]
[puts_can_be_seen_on_other_clients_after_failover(org.terracotta.management.integration.tests.FailoverIT)][testServer0] 2016-12-19 16:12:09,340 [WorkerThread(l2_state_change_stage, 0)] ERROR com.tc.server.TCServerMain - Thread:Thread[WorkerThread(l2_state_change_stage, 0),5,TC Thread Group] got an uncaught exception. calling CallbackOnExitDefaultHandlers.
[puts_can_be_seen_on_other_clients_after_failover(org.terracotta.management.integration.tests.FailoverIT)][testServer0] com.tc.exception.TCRuntimeException: Uncaught exception in stage
[puts_can_be_seen_on_other_clients_after_failover(org.terracotta.management.integration.tests.FailoverIT)][testServer0] 	at com.tc.async.impl.StageImpl$WorkerThread.run(StageImpl.java:211)
[puts_can_be_seen_on_other_clients_after_failover(org.terracotta.management.integration.tests.FailoverIT)][testServer0] Caused by: java.lang.IllegalStateException: Server testServer0 is not active!
```

we should have instead some M&M traces like these:

```
2016-12-16 18:17:05,462 [WorkerThread(l2_election_handler_stage, 0)] INFO com.tc.l2.state.StateManagerImpl - Becoming State[ ACTIVE-COORDINATOR ]
2016-12-16 18:17:05,463 [WorkerThread(l2_election_handler_stage, 0)] INFO com.terracottatech.console - Becoming State[ ACTIVE-COORDINATOR ]
2016-12-16 18:17:05,463 [WorkerThread(l2_election_handler_stage, 0)] INFO com.tc.l2.state.ElectionManagerImpl - Declared as Winner: Winner is : Enrollment [ NodeID[localhost:22144], isNew = false, weights = 28,0,7612,-6858803596795605944 ]
2016-12-16 18:17:05,469 [WorkerThread(l2_state_change_stage, 0)] TRACE org.terracotta.management.service.monitoring.IStripeMonitoringPlatformListenerAdapter - [0] serverDidBecomeActive(testServer1)
2016-12-16 18:17:05,469 [WorkerThread(l2_state_change_stage, 0)] TRACE org.terracotta.management.service.monitoring.TopologyService - [0] serverDidBecomeActive(testServer1)
2016-12-16 18:17:05,469 [WorkerThread(l2_state_change_stage, 0)] TRACE org.terracotta.management.service.monitoring.TopologyService - [0] serverDidJoinStripe(testServer1)
```
And then, after the `serverDidBeconeActive` call, we should have the replay phase, and then the addNode() when the active entities are created.

[server0.stdout.log.txt](https://github.com/Terracotta-OSS/terracotta-core/files/662283/server0.stdout.log.txt)

[server1.stdout.log.txt](https://github.com/Terracotta-OSS/terracotta-core/files/662284/server1.stdout.log.txt)

I am using:

```
    <terracotta-core.version>5.2-SNAPSHOT</terracotta-core.version>
    <galvan.version>1.2-SNAPSHOT</galvan.version>
```

**Comment by mathieucarbou** [comment]
cc @anthonydahanne

**Comment by myronkscott** [comment]
Looking at the code here

https://github.com/mathieucarbou/terracotta-platform/blob/ittests/management/monitoring-service/src/main/java/org/terracotta/management/service/monitoring/MonitoringServiceProvider.java#L183

It looks like this is a simple assert on state.  We have discussed this before but entity state will not exactly track server state.  The server is in a state of transition.  Once all the entities have been transitioned, the server will be transitioned.  Normally, services will not have notion of server state.  Monitoring is special in it gets these events from the platform directly.  I'm not sure that the service should be using this information directly.  

So, is it possible to remove this assert which the knowledge that creating active entities can happen before the server transitions  (failover case) and after (new creates).  

We can fake the series of events but my fear is that rearranging the events like this from the platform will involve a good deal of fakery that will be difficult to maintain.

**Comment by mathieucarbou** [comment]
It's not an error in this code @myronkscott : we have some guards here because at this point, if an entity is requesting such service, it means the monitoring tree SHOULD be up and ready. 

__Monitoring services cannot work if the platform did not sent the right events__ because we do not have the right topology in place.

The code is right. It fails upfront, before any entity can do any potential harmful work. If we do not have this guard it will fail later, because platform did not provided the right information before.

I won't change that, because it's right to have this check here. We cannot do anything in monitoring service BEFORE server sent us the __serviceDidBecomeActive()__ event.

Any entity trying to do anything will fail. So this is a guard, and it works fine because it showed a platform bug!

**Comment by mathieucarbou** [comment]
Also, if we use the released version, the code works fine: the topology gets constructed. In the snapshot, it doesn't work. So there has been a change impacted this.

The ordering has always been the same:

1. Event serverDidBecomeActive()
2. Event bout client connection / entity creation
3. Events about fetch

I can provide you some logs if you need to see in an active / passive case (not failover) what is the right ordering, in the current released versions we use in tc-platform.

**Comment by mathieucarbou** [comment]
@myronkscott please have a look at _server1.stdout.log.txt_ attached above. You'll see several logs from `IStripeMonitoringPlatformListenerAdapter`. This is the implementation of the `IStripeMonitoring` interface.

Here, we should be informed about __ALL__ platform events: when a server state changes, server become active, server comes, leaves, entities get created, client connects or disconnect, etc.

M&M services relies on this information to work. And we should be aware of __ALL__ platform events thus.

**Comment by myronkscott** [comment]
You are not missing events.  You are getting them in different order.  My feeling is that maybe you are too reliant on message ordering which will be incredibly brittle for both of us to maintain.  I will discuss with Jeff and get back to you.

**Comment by mathieucarbou** [comment]
It's not true I think because if we look at the logs here:

```
2016-12-19 16:12:09,330 [WorkerThread(l2_election_handler_stage, 0)] INFO com.tc.l2.state.StateManagerImpl - Becoming State[ ACTIVE-COORDINATOR ]
2016-12-19 16:12:09,330 [WorkerThread(l2_election_handler_stage, 0)] INFO com.terracottatech.console - Becoming State[ ACTIVE-COORDINATOR ]
2016-12-19 16:12:09,331 [WorkerThread(l2_election_handler_stage, 0)] INFO com.tc.l2.state.ElectionManagerImpl - Declared as Winner: Winner is : Enrollment [ NodeID[localhost:22965], isNew = false, weights = 11,0,1968,-6114697689862609498 ]
2016-12-19 16:12:09,340 [WorkerThread(l2_state_change_stage, 0)] ERROR com.tc.server.TCServerMain - Thread:Thread[WorkerThread(l2_state_change_stage, 0),5,TC Thread Group] got an uncaught exception. calling CallbackOnExitDefaultHandlers.
com.tc.exception.TCRuntimeException: Uncaught exception in stage
```

It says that the server did become active, and then an active entity is created.

We expect to have an event `serverDidBecomeActive()` before entity creation, and an event telling us about entity creation after the entity gets created.

Why changing some behaviors that have been discussed months ago already ? All the monitoring code relies on the fact that we receive platform events as things occur. When an entity is created, it starts using monitoring services. Monitoring services cannot be used if the topology is not there because we do not have the information we need to create the management objects required for TMS.

**Comment by myronkscott** [comment]
I also told you months ago that entity state does not exactly track server state and to not rely on them to always be perfectly matched.  I am actively looking at this but you do not have the full scope of issues to dictate the fix for this.  Please be patient while I look for a proper solution.

**Comment by mathieucarbou** [comment]
FYI, just to confirm, I tried to remove this guards, and it fails in the ehcache management code because the topology has not yet been constructed, because we missed some platform events.

**Comment by myronkscott** [comment]
This is the platform's order of events currently

1.  Election manager declares winner (ACTIVE-COORINDATOR logging that you see)
but the server is not fully active yet.  That log message only starts the process
2.  Passive entities are removed
3.  Active entities are created
4.  Server declared active to M&M and cached tree replayed
5.  Reconnect window opens
6.  Clients reconnect
7.  Reconnect window closes transition completed to active server is now ready to process new transactions

So in step 3, please help me understand what the active entity is expecting to see that I suspect it will get in step 4.  This change was made because the previous behavior was racy, reporting from multiple threads with the possibility of getting double events or events interleaved in a non-deterministic way.

**Comment by mathieucarbou** [comment]
@myronkscott @jd0-sag @anthonydahanne :

I tested removing the "guards" and typically allow the monitoring producer to:

- either provide the (same) service instance, but shared between a passive and an active at the same time (because same consumer id)
- either provide 1 service instance per entity and keep track of these creation (at most 2 per consumer id)
- either, if i find an existing service for the consumer id, i am clearing the states, and re-create a new one
This would solve one of the issue, and allows to go further.

So these changes works. So depending on the services, either I keep track on an active or passive version, or I am reusing it, or I am recreating it after cleaning up the previous states.

But if I look at the sequence I have, this is not what you gave me above:

5. Server declared active to M&M and cached tree replayed

=> actually I only have the addNode() with isActive=true, for active entity addition. I do not have any replay at all. I am losing the fact that there has been a failover (that I was able to detect with an addNode(entity A, isActive=false), followed by an addNode(entity A, isActive=true))

See the logs:

[server0.stdout.log.txt](https://github.com/Terracotta-OSS/terracotta-core/files/664645/server0.stdout.log.txt)
[server1.stdout.log.txt](https://github.com/Terracotta-OSS/terracotta-core/files/664644/server1.stdout.log.txt)

So with this change plus my changes, currently, we go further but i am not able to detect the failover since I do not have the 2 addNode calls with isActive=false and iActive=true
